File: 128_LongestConsecutiveSequence.cpp
// Topic: Longest Consecutive Sequence
// Description: Given an unsorted array of integers, find the length of the longest consecutive
// elements sequence. Approach: Use an unordered_set for O(1) lookup. For each number that is the
// start of a sequence (num-1 not in set), count consecutive numbers.

#include <bits/stdc++.h>
using namespace std;

int longestConsecutive(vector<int>& nums)
{
    unordered_set<int> s(nums.begin(), nums.end());
    int best = 0;
    for (int num : s)
    {
        if (s.find(num - 1) != s.end())
            continue;  // not start
        int cur = num, len = 1;
        while (s.find(cur + 1) != s.end())
        {
            cur++;
            len++;
        }
        best = max(best, len);
    }
    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<int>> tests = {{100, 4, 200, 1, 3, 2}, {0, 3, 7, 2, 5, 8, 4, 6, 0, 1}};
    for (int i = 0; i < (int)tests.size(); ++i)
        cout << "Test " << (i + 1) << ": " << longestConsecutive(tests[i]) << '\n';
    return 0;
}

/*
Time Complexity: O(n) average
Space Complexity: O(n)
*/

File: 128_LongestConsecutiveSequence.py
# Topic: Longest Consecutive Sequence
# Description: Given an unsorted array of integers, find the length of the longest consecutive elements sequence.
# Approach: Use a set for O(1) lookup. For each number that is the start of a sequence (num-1 not in set),
#          count consecutive numbers.

from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        num_set = set(nums)
        best = 0
        for num in num_set:
            if num - 1 in num_set:
                continue  # not start of sequence
            cur = num
            length = 1
            while cur + 1 in num_set:
                cur += 1
                length += 1
            best = max(best, length)
        return best


if __name__ == "__main__":
    tests = [[100, 4, 200, 1, 3, 2], [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]]
    for i, nums in enumerate(tests):
        print(f"Test {i + 1}: {Solution().longestConsecutive(nums)}")

"""
Time Complexity: O(n) average
Space Complexity: O(n)
"""
